### DISCUSSION TITLE
How to use Avalon to implement a virtual keyboard and simulate user input functions?

### DISCUSSION THREAD

**Discussion by susin3x** [AvaloniaUI/Avalonia#17632] [Q&A]
I want to use Avalonia to implement virtual keyboard functionality and simulate user input. Do you have any recommended cross platform libraries? Which one is more recommended, using cross platform libraries and conditional compilation or runtime checks to call input libraries from different platforms?

**Comment by MikeCodesDotNET** [comment]
As part of [Avalonia Accelerate](https://github.com/AvaloniaUI/Avalonia/discussions/16997), we're implementing a virtual keyboard that may suit your needs.

**Comment by thevortexcloud** [comment]
The keyboard itself is not that hard. It's just a bunch of buttons. You can have a XAML file per layout if you want to support multiple layouts. I wrote one awhile back that ended up looking like this for a US QWERTY layout:

```xaml
            <StackPanel Orientation="Horizontal">
                <views:VirtualKeyboardKey LowerKey='`' UpperKey="~" />
                <views:VirtualKeyboardKey LowerKey='1' UpperKey="!" />
                <views:VirtualKeyboardKey LowerKey='2' UpperKey="@" />
                <views:VirtualKeyboardKey LowerKey='3' UpperKey="#" />
                <views:VirtualKeyboardKey LowerKey='4' UpperKey="$" />
                <views:VirtualKeyboardKey LowerKey='5' UpperKey="%" />
                <views:VirtualKeyboardKey LowerKey='6' UpperKey="^" />
                <views:VirtualKeyboardKey LowerKey='7' UpperKey="&amp;" />
                <views:VirtualKeyboardKey LowerKey='8' UpperKey="*" />
                <views:VirtualKeyboardKey LowerKey='9' UpperKey="(" />
                <views:VirtualKeyboardKey LowerKey='0' UpperKey=")" />
                <views:VirtualKeyboardKey LowerKey='-' UpperKey="_" />
                <views:VirtualKeyboardKey LowerKey='=' UpperKey="+" />
                <views:VirtualKeyboardKey Mode="Special" SpecialKeyType="Backspace" Content="Backspace" />
            </StackPanel>
```
Then just repeat for every row of the keyboard and do some handling for when a key is prssed.

However my keyboard was only usable within Avalonia itself (and even then, only within text boxes). Making one that works outside Avalonia is a lot trickier. On Linux, your easiest approach is probably to just create a userspace keyboard driver that runs when your app is up. There is an example in the Kernel docs for C on how to do that.

https://www.kernel.org/doc/html/v6.11/input/uinput.html

If you do that, then as far as anything else is concerned you have a proper keyboard connected. I can't really comment on other platforms. One part I am not sure about is how you would press a key without losing focus on what you are trying to type into. 

> using cross platform libraries and conditional compilation or runtime checks to call input libraries from different platforms?

I personally prefer runtime checks purely because it makes development significantly less painful with cross compiling. But if you can use an existing library then you should. That said C# itself does not care at compile time if you reference a platform specific nativelibrary on a platform it does not exist for.
